File: java/Lexar.java

class Lexar {
  public String input = "";

  private boolean returnPreviousToken = false;
  private Token previousToken;

  public Lexar(String input) {
    this.input = input.trim();
  }

  public Token getNextToken() {
    if (this.returnPreviousToken == true) {
      this.returnPreviousToken = false;
      return this.previousToken;
    }

    Token token = new Token();
  
    char c = getNextChar();

    switch (c) {
      case 'a': token.setKind(Token.PLUS); break;
      case 'b': token.setKind(Token.MINUS); break;
      case 'c': token.setKind(Token.MULTIPLY); break;
      case 'd': token.setKind(Token.DIVIDE); break;
      case 'e': token.setKind(Token.LPAREN); break;
      case 'f': token.setKind(Token.RPAREN); break;
      case '\0': token.setKind(Token.EXPR_END); break;
      default:
        token.setKind(Token.NUMBER); 
        String value = String.valueOf(c);

        while (this.input.matches("^[0-9].*")) {
          value = value + getNextChar();
        }
        token.setValue(value);
    }

    this.previousToken = token;
    return token;
  }

  public void revert() {
    this.returnPreviousToken = true;
  }

  private char getNextChar() {
    char c = '\0'; // Empty char

    if (this.input.length() > 0) {
      c = this.input.charAt(0);
      this.input = this.input.substring(1);
    }
    return c;
  }
}

File: ruby/lexar.rb
require_relative 'token'

class Lexar
  def initialize(input)
    @input = input
    @return_previous_token = false
  end

  def get_next_token
    if @return_previous_token
      @return_previous_token = false
      return @previous_token
    end

    token = Token.new

    @input.lstrip!

    # Operator mapping
    # a = +
    # b = -
    # c = *
    # d = /
    # e = (
    # f = )

    case @input
      when /\Aa/ then
        token.kind = Token::PLUS
      when /\Ab/ then
        token.kind = Token::MINUS
      when /\Ac/ then
        token.kind = Token::MULTIPLY
      when /\Ad/ then
        token.kind = Token::DIVIDE
      when /\A\d+(\.\d+)?/
        token.kind = Token::NUMBER
        token.value = $&.to_f
      when /\Ae/
        token.kind = Token::LPAREN
      when /\Af/
        token.kind = Token::RPAREN
      when ''
        token.kind = Token::EXPR_END
    end

    raise 'Unknown token' if token.unknown?
    @input = $'

    @previous_token = token
    token
  end

  def revert
    @return_previous_token = true
  end
end
